pkg/psmdb/mongos.go

```diff
@@ -0,0 +1,361 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func MongosDeployment(cr *api.PerconaServerMongoDB) *appsv1.Deployment {
+	return &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "apps/v1",
+			Kind:       "Deployment",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-" + "mongos",
+			Namespace: cr.Namespace,
+		},
+	}
+}
+
+func MongosDeploymentSpec(cr *api.PerconaServerMongoDB, operatorPod corev1.Pod) (appsv1.DeploymentSpec, error) {
+	ls := map[string]string{
+		"app.kubernetes.io/component": "mongos",
+		"app.kubernetes.io/instance":  cr.Name,
+	}
+
+	c, err := mongosContainer(cr)
+	if err != nil {
+		return appsv1.DeploymentSpec{}, fmt.Errorf("failed to create container %v", err)
+	}
+
+	initContainers := initContainers(cr, operatorPod)
+	for i := range initContainers {
+		initContainers[i].Resources.Limits = c.Resources.Limits
+		initContainers[i].Resources.Requests = c.Resources.Requests
+	}
+
+	return appsv1.DeploymentSpec{
+		Replicas: &cr.Spec.Sharding.Mongos.Size,
+		Selector: &metav1.LabelSelector{
+			MatchLabels: ls,
+		},
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels:      ls,
+				Annotations: cr.Spec.Sharding.Mongos.MultiAZ.Annotations,
+			},
+			Spec: corev1.PodSpec{
+				SecurityContext:   cr.Spec.Sharding.Mongos.PodSecurityContext,
+				Affinity:          PodAffinity(cr.Spec.Sharding.Mongos.MultiAZ.Affinity, ls),
+				NodeSelector:      cr.Spec.Sharding.Mongos.MultiAZ.NodeSelector,
+				Tolerations:       cr.Spec.Sharding.Mongos.MultiAZ.Tolerations,
+				PriorityClassName: cr.Spec.Sharding.Mongos.MultiAZ.PriorityClassName,
+				RestartPolicy:     corev1.RestartPolicyAlways,
+				ImagePullSecrets:  cr.Spec.ImagePullSecrets,
+				Containers:        []corev1.Container{c},
+				InitContainers:    initContainers,
+				Volumes:           volumes(cr),
+				SchedulerName:     cr.Spec.SchedulerName,
+			},
+		},
+	}, nil
+}
+
+func initContainers(cr *api.PerconaServerMongoDB, operatorPod corev1.Pod) []corev1.Container {
+	inits := []corev1.Container{}
+	if cr.CompareVersion("1.5.0") >= 0 {
+		inits = append(inits, EntrypointInitContainer(operatorPod.Spec.Containers[0].Image))
+	}
+
+	return inits
+}
+
+func mongosContainer(cr *api.PerconaServerMongoDB) (corev1.Container, error) {
+	fvar := false
+
+	resources, err := CreateResources(cr.Spec.Sharding.Mongos.ResourcesSpec)
+	if err != nil {
+		return corev1.Container{}, fmt.Errorf("resource creation: %v", err)
+	}
+
+	volumes := []corev1.VolumeMount{
+		{
+			Name:      MongodDataVolClaimName,
+			MountPath: MongodContainerDataDir,
+		},
+		{
+			Name:      InternalKey(cr),
+			MountPath: mongodSecretsDir,
+			ReadOnly:  true,
+		},
+		{
+			Name:      "ssl",
+			MountPath: sslDir,
+			ReadOnly:  true,
+		},
+		{
+			Name:      "ssl-internal",
+			MountPath: sslInternalDir,
+			ReadOnly:  true,
+		},
+	}
+
+	if *cr.Spec.Mongod.Security.EnableEncryption {
```

Comment:
why it is needed for mongos?

Reply:
Do not needed
https://github.com/percona/percona-server-mongodb-operator/pull/439/commits/be7bdefaa2b6f362495a8a77e1b79e5e50ad2560